### DISCUSSION TITLE
Dynamic loading for multiApps

### DISCUSSION THREAD

**Discussion by rtaylo45** [idaholab/moose#18217] [Q&A General] [answered]
Hello,

I am running into an issue when trying to run multiapps. I have my main app called mole and the subapp called SAM. SAM uses git submodule for MOOSE and has it located in its main directory. I have set my `MOOSE_DIR` path for both SAM and mole to point to the git submodule for MOOSE in SAM. I have built both codes using this MOOSE directory. Both codes compile and are able to run independently. I have a simple input file which uses the mutliapps block to call SAM from mole. 

```
[MultiApps]
  [sam]
    type = FullSolveMultiApp
    app_type = SAMApp
    input_files = 'sam_pipe.i'
    execute_on = 'INITIAL'
  []
[]
```

I have tried two ways of setting the library path. First by setting `library_path` in the MultiApps block to the SAM lib. Second by exporting the `MOOSE_LIBRARY_PATH` to the SAM lib folder. Both of these methods lead to the same error, shown below.

```
*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/tensor_mechanics/lib/libtensor_mechanics-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/ray_tracing/lib/libray_tracing-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/heat_conduction/lib/libheat_conduction-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/ray_tracing/lib/libray_tracing-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/fluid_properties/lib/libfluid_properties-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/module_loader/lib/libmodule_loader_with_fp_ray_hc_tm-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/tensor_mechanics/lib/libtensor_mechanics-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/ray_tracing/lib/libray_tracing-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/heat_conduction/lib/libheat_conduction-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/ray_tracing/lib/libray_tracing-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/moose/modules/fluid_properties/lib/libfluid_properties-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** Warning ***
Unable to find extern "C" method "SAMApp__registerApps" in library: /Users/qt5/Desktop/coding/projects/SAM/lib/libsam-dbg.0.dylib.
This doesn't necessarily indicate an error condition unless you believe that the method should exist in that library.

*** ERROR ***
Unable to locate library for "SAMApp".
We attempted to locate the library "libsam-dbg.la" in the following paths:

Make sure you have compiled the library and either set the "library_path" variable in your input file or exported "MOOSE_LIBRARY_PATH".
Compiled in debug mode to see the list of libraries checked for dynamic loading methods.```

**Comment by rwcarlsen** [comment, marked as answer]
From those warnings, it is clear that libsam is being correctly discovered and searched.  But there should be an extern C (exported) function like this: https://github.com/idaholab/moose/blob/d4a83b1aa02b896d00b6ef07dcc82efeedf38b52/test/src/base/MooseTestApp.C#L125.  Can you confirm such functions exist for SamApp?

**Reply by rtaylo45** [reply]
Im not familiar with these lines of code from that file you mentioned. 

```
extern "C" void
MooseTestApp__registerAll(Factory & f, ActionFactory & af, Syntax & s)
{
  MooseTestApp::registerAll(f, af, s);
}

// External entry point for dynamic application loading
extern "C" void
MooseTestApp__registerApps()
{
  MooseTestApp::registerApps();
}
```

Ive never see it in any documentation or in the MOOSE tutorials. Can you provide more information on them?

**Reply by rwcarlsen** [reply]
We don't say too much about them because the stork templates we provide for people to create new apps with automatically include/provide these functions.  Users generally should never need to do anything with them.  But it's possible these functions were inadvertently removed from SAM.  If there are not similar functions in SAM, then this pattern will need to be followed to add them back (in the SAMApp.C file).  i.e. add back `SAMApp__registerApps() {SAMApp::registerApps();}` and `SAMApp__registerAll(...) {SAMApp::registerAll(...);}`

**Reply by rtaylo45** [reply]
These functions are in SAMApp.C

**Reply by rwcarlsen** [reply]
Hmmm...  That's too bad because that would have been an easy fix.  Maybe @permcody has some ideas.  I know that there have been some issues related to dynamic library loading for us on macs.

**Reply by rtaylo45** [reply]
I am running on a mac, Big Sur version 11.4. What issues are there on the macs?

**Reply by rwcarlsen** [reply]
Your issue doesn't seem to be the same as some of the other ones, but if you just search our issues for "dynamic loading" you'll find several.  Some to do with path handling, some to do with inter-module dependencies, some to do with crashes, etc.  Have you tried running your scenario/setup on a linux system?  Does it work there or do you observe the same problem?

**Reply by rtaylo45** [reply]
So i do get this same error when i try to run on linux.

**Reply by andrsd** [reply]
Your error message is complaining about `SAMApp__registerApps`, but SAM implemented `SamApp__registerApps`, `SamApp__registerAll` and `SamApp__registerObjects`.

Probably you want to change `app_type = SAMApp` to `app_type = SamApp`. These things are case-sensitive.

**Reply by rtaylo45** [reply]
Wow, i feel dumb now. This fixed the issue. Thank you for the help.

**Reply by andrsd** [reply]
Cool! Glad to hear that. Maybe, ping the SAM developers to they make spelling of SAM consistent thought out the code. Sometimes that use `SAM`, sometimes they use `Sam`. Then, it leads to problems like this.
